### PR TITLE
Fully setup kernelspecs refs

### PR DESCRIPTION
**Pull request by theengineear**
This was a pretty easy spot to start. It's not really complete in that `hostRef` isn't filled in since we don't have the notion of `hostRef` in the apps yet (next up). But all the actions / actionTypes were already there.

I'd love to discuss some of the organizational bits of this before I pile on more and more PRs like this :)

screen shot 2018-02-20 at 7 47 23 pm

### CODE REVIEW

File: packages/core/plan.md [hunk 1]
```diff
@@ -181,7 +181,8 @@ type core = {
             }
           },
         }
-      }
+      },
+      refs: List<KernelspecsRef>
```
Comment (theengineear) [inline packages/core/plan.md:185]: We're going to want to map over this in the menu, right?

File: applications/jupyter-extension/nteract_on_jupyter/app/store.js [hunk 2]
```diff
@@ -50,12 +49,7 @@ const rootReducer = combineReducers({
   comms,
   config,
   modals,
-  communication: combineReducers({
-    kernelspecsByRef: communicationKernelspecsByRef
-  }),
-  entities: combineReducers({
-    kernelspecsByRef: entitiesKernelspecsByRef
-  })
```
Comment (theengineear) [inline applications/jupyter-extension/nteract_on_jupyter/app/store.js:58]: I'm pretty sure these weren't being used yet.

File: applications/jupyter-extension/nteract_on_jupyter/app/store.js [hunk 3]
```diff
@@ -50,12 +49,7 @@ const rootReducer = combineReducers({
...
+  core
```
Comment (theengineear) [inline applications/jupyter-extension/nteract_on_jupyter/app/store.js:52]: `core` is our main point of entry for core state moving forward.

File: packages/core/src/selectors/index.js [hunk 4]
```diff
@@ -242,3 +242,18 @@ export const currentTheme: (state: *) => string = createSelector(
   (state: AppState) => state.config.get("theme", "light"),
   identity
 );
+
+// TODO: These selectors all have "Core" suffixes that will be removed when we
+// fully switch over to using the new state tree under state.core.
+
+export const communicationKernelspecsByRefCore = createSelector(
+  (state: AppState, { kernelspecsRef }) =>
```
Comment (theengineear) [inline packages/core/src/selectors/index.js:250]: @rgbkrk this is more or less what you were thinking for our byRef-y selectors, right?

File: packages/core/src/types/state/ids.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+// ... where 9092-7679-9978-8822 is the KernelId.
+//
+// We call it "castTo*" to make it clear that you're not getting anything unique
+// out of the function.
+
+export opaque type HostId = string;
+export opaque type KernelId = string;
+
+export const castToHostId = (id: string): HostId => id;
+export const castToKernelId = (id: string): KernelId => id;
```
Comment (theengineear) [inline packages/core/src/types/state/ids.js:14]: Not being used yet, but I snuck these in...

File: packages/core/src/types/state/refs.js [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+// See https://flow.org/en/docs/types/opaque-types/#toc-within-the-defining-file
+export opaque type HostRef = string;
+export opaque type KernelRef = string;
+export opaque type KernelspecsRef = string;
+export opaque type ContentRef = string;
+
+export const createHostRef = (): HostRef => uuid.v4();
+export const createKernelRef = (): KernelRef => uuid.v4();
+export const createKernelspecsRef = (): KernelspecsRef => uuid.v4();
+export const createContentRef = (): ContentRef => uuid.v4();
```
Comment (theengineear) [inline packages/core/src/types/state/refs.js:19]: ^^ Not all being used yet, but I snuck these in.

**Comment by theengineear** [comment]
Ooh! Actually, I will make one more commit to start using the _real_ `KernelRef` types instead of that hard-coded string...

screen shot 2018-02-20 at 8 00 47 pm

File: applications/jupyter-extension/nteract_on_jupyter/app/store.js (hunk 3 above)
Comment (rgbkrk) [reply applications/jupyter-extension/nteract_on_jupyter/app/store.js:52]: Woo!

File: applications/jupyter-extension/nteract_on_jupyter/app/index.js [hunk 7]
```diff
@@ -27,14 +29,10 @@ function createApp(jupyterConfigData: JupyterConfigData) {
   class App extends React.Component<*> {
     notificationSystem: NotificationSystem;
 
-    // TODO: the kernelspecsRef is hard-coded to be 'single-server' in this
-    // application because we only anticipate _one_ set of possible kernelspecs.
-    // However, since `/core` assumes that a generic notebook application may
-    // be able to connect to multiple servers and thus have many kernelspecs,
-    // it needs a ref to complete the action.
     componentDidMount(): void {
+      const kernelspecsRef = createKernelspecsRef();
```
Comment (rgbkrk) [inline applications/jupyter-extension/nteract_on_jupyter/app/index.js:33]: 👍

File: packages/core/plan.md (hunk 1 above)
Comment (rgbkrk) [reply packages/core/plan.md:185]: Yup!

Comment (rgbkrk) [reply packages/core/plan.md:185]: This makes me realize something. A host only has one set of kernelspecs.

File: packages/core/src/reducers/index.js [hunk 8]
```diff
@@ -4,17 +4,6 @@ import comms from "./comms";
...
-export {
-  document,
-  comms,
-  config,
-  app,
-  modals,
-  communicationKernelspecsByRef,
-  entitiesKernelspecsByRef
-};
+export { document, comms, config, app, modals, core };
```
Comment (rgbkrk) [inline packages/core/src/reducers/index.js:9]: 👍

File: packages/core/src/selectors/index.js (hunk 4 above)
Comment (rgbkrk) [reply packages/core/src/selectors/index.js:250]: Yup, props to get the arbitrary keyed elements.

File: packages/core/src/types/state/refs.js (hunk 6 above)
Comment (rgbkrk) [reply packages/core/src/types/state/refs.js:19]: Works for me, might as well have them situated since they're part of the core plan.

**Review by rgbkrk — APPROVED** [review]

**Pull request merged**